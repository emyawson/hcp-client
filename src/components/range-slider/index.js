export * from './range-slider.style';
export * from './range-slider.component';
