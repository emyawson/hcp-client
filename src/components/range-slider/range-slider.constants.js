export const HANDLES = {
  MAX_HANDLE: 'maxHandle',
  MIN_HANDLE: 'minHandle',
};
