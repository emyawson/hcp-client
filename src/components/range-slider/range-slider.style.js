import styled from 'styled-components';

import { colors } from 'src/core/styles/colors';

const handleHeight = 20;
const handleWidth = 20;
const pathHeight = 6;

export const RangeSliderWrapper = styled.div`
  height: 4rem;
  position: relative;
  z-index: 1;
  user-select: none;
  display: flex;
`;

RangeSliderWrapper.displayName = 'RangeSliderWrapper';

export const Handle = styled.div`
  position: absolute;
  top: -${handleHeight / 2 - pathHeight / 2}px;
  left: ${props => (props.left || 0) - handleWidth / 2}px;
  z-index: ${props => (props.dragging ? 4 : 3)};
  width: ${handleHeight / 16}rem;
  height: ${handleHeight / 16}rem;
  border-radius: 50%;
  background-color: ${colors.white};
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.5);
  cursor: ew-resize;
`;

Handle.displayName = 'Handle';

export const Path = styled.div`
  position: relative;
  z-index: 1;
  height: ${pathHeight / 16}rem;
  border-radius: ${pathHeight}rem;
  background-color: ${colors.blueFaded};
  margin: 0 auto;
  flex-grow: 1;
`;

Path.displayName = 'Path';

export const InnerPath = styled.div`
  position: absolute;
  z-index: 1;
  height: ${pathHeight / 16}rem;
  background-color: ${colors.blue};
  top: 0;
  left: ${props => props.left || 0}px;
  width: ${props => props.width || 0}px;
`;

InnerPath.displayName = 'InnerPath';
