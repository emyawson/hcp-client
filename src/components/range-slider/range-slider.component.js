import React, { Component } from 'react';

import { Path, Handle, InnerPath } from './range-slider.style';
import { HANDLES } from './range-slider.constants';
import { getHandleState } from './range-slider.utils';

export class RangeSlider extends Component {
  state = {
    draggingHandle: null,
    initialClientX: 0,
    maxHandleValue: 0,
    minHandleValue: 0,
    rangeMax: 0,
    rangeMin: 0,
    tickDistance: 0,
  };

  componentDidMount() {
    this.initialize(this.props);
    document.addEventListener('mousemove', this.onMouseMove);
    document.addEventListener('mouseup', this.onMouseUp);
    window.addEventListener('resize', this.adaptRangeSliderToWindowSize);
  }

  componentWillReceiveProps(nextProps) {
    this.initialize(nextProps);
  }

  componentWillUnmount() {
    document.removeEventListener('mousemove', this.onMouseMove);
    document.removeEventListener('mouseup', this.onMouseUp);
    window.removeEventListener('resize', this.adaptRangeSliderToWindowSize);
  }

  render() {
    const {
      rangeMin,
      minHandleValue,
      maxHandleValue,
      tickDistance,
    } = this.state;
    return (
      <Path innerRef={ref => (this.path = ref)}>
        <InnerPath
          left={(minHandleValue - rangeMin) * tickDistance}
          width={(maxHandleValue - minHandleValue) * tickDistance}
        />
        <Handle
          left={(minHandleValue - rangeMin) * tickDistance}
          onMouseDown={this.onMouseDown(HANDLES.MIN_HANDLE)}
          innerRef={ref => (this.minHandle = ref)}
        />
        <Handle
          left={(maxHandleValue - rangeMin) * tickDistance}
          onMouseDown={this.onMouseDown(HANDLES.MAX_HANDLE)}
          innerRef={ref => (this.maxHandle = ref)}
        />
      </Path>
    );
  }

  adaptRangeSliderToWindowSize = e => this.initialize(this.props);

  initialize = props => {
    const { minValue, maxValue, rangeMin, rangeMax } = props;
    const range = rangeMax - rangeMin;
    const tickDistance = this.path && this.path.clientWidth / range;
    this.setState({
      maxHandleValue: maxValue,
      minHandleValue: minValue,
      rangeMax,
      rangeMin,
      tickDistance,
    });
  };

  onMouseDown = handleId => e =>
    this.setState({ draggingHandle: handleId, initialClientX: e.clientX });

  onMouseMove = e => {
    const {
      draggingHandle,
      maxHandleValue,
      minHandleValue,
      rangeMin,
      rangeMax,
      tickDistance,
    } = this.state;
    if (draggingHandle) {
      const { path } = this;
      const pathX = e.clientX - path.getBoundingClientRect().left;
      const pathWidth = path.clientWidth;
      const nextState = getHandleState(
        pathWidth,
        pathX,
        tickDistance,
        draggingHandle,
        minHandleValue,
        maxHandleValue,
        rangeMin,
        rangeMax,
      );
      this.setState(nextState);
    }
  };

  onMouseUp = e => {
    if (this.state.draggingHandle) {
      this.setState({ draggingHandle: null, initialClientX: 0 });
      this.handleOnChange();
    }
  };

  handleOnChange = () => {
    const { minHandleValue, maxHandleValue } = this.state;
    this.props.onChange({ max: maxHandleValue, min: minHandleValue });
  };
}
RangeSlider.defaultProps = {
  maxValue: 1,
  minValue: 0,
  onChange: () => undefined,
  rangeMax: 1,
  rangeMin: 0,
};
