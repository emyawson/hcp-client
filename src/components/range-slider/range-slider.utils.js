import { HANDLES } from './range-slider.constants';

export const getHandleDragPosition = (
  pathWidth,
  pathX,
  tickDistance,
  rangeMin,
  rangeMax,
) => {
  if (pathX < 0) {
    return rangeMin;
  } else if (pathX > pathWidth) {
    return rangeMax;
  } else {
    return Math.ceil(pathX / tickDistance) + rangeMin;
  }
};

export const getHandleState = (
  pathWidth,
  pathX,
  tickDistance,
  draggingHandle,
  minHandleValue,
  maxHandleValue,
  rangeMin,
  rangeMax,
) => {
  let nextStartHandleValue =
    draggingHandle === HANDLES.MIN_HANDLE
      ? getHandleDragPosition(
          pathWidth,
          pathX,
          tickDistance,
          rangeMin,
          rangeMax,
        )
      : minHandleValue;
  let nextEndHandleValue =
    draggingHandle === HANDLES.MAX_HANDLE
      ? getHandleDragPosition(
          pathWidth,
          pathX,
          tickDistance,
          rangeMin,
          rangeMax,
        )
      : maxHandleValue;

  if (nextStartHandleValue > nextEndHandleValue) {
    nextStartHandleValue = minHandleValue;
    nextEndHandleValue = maxHandleValue;
  }

  return {
    maxHandleValue: nextEndHandleValue,
    minHandleValue: nextStartHandleValue,
  };
};
