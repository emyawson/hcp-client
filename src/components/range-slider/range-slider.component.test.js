import React from 'react';
import { shallow } from 'enzyme';

import { RangeSlider } from './range-slider.component';

describe('RangeSlider Component Tests', () => {
  const mockOnChange = jest.fn().mockName('onRangeSliderValuesChanged');
  const mockProps = {
    minValue: 3,
    maxValue: 46,
    rangeMin: 0,
    rangeMax: 50,
    onChange: mockOnChange,
  };
  it('Should render correctly', () => {
    const wrapper = shallow(<RangeSlider {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('Should dispatch an event when user has changed values', () => {
    const wrapper = shallow(<RangeSlider {...mockProps} />);
    wrapper.setState({ minHandleValue: 5, maxHandleValue: 10 });
    wrapper.instance().handleOnChange();
    expect(mockOnChange).toHaveBeenCalled();
    expect(mockOnChange).toHaveBeenCalledWith({ min: 5, max: 10 });
  });
});
