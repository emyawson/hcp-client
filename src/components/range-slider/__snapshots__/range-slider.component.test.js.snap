// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeSlider Component Tests Should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RangeSlider
    maxValue={46}
    minValue={3}
    onChange={[MockFunction onRangeSliderValuesChanged]}
    rangeMax={50}
    rangeMin={0}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <InnerPath
          left={NaN}
          width={NaN}
        />,
        <Handle
          innerRef={[Function]}
          left={NaN}
          onMouseDown={[Function]}
        />,
        <Handle
          innerRef={[Function]}
          left={NaN}
          onMouseDown={[Function]}
        />,
      ],
      "innerRef": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {
          "left": NaN,
          "width": NaN,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {
          "innerRef": [Function],
          "left": NaN,
          "onMouseDown": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {
          "innerRef": [Function],
          "left": NaN,
          "onMouseDown": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <InnerPath
            left={NaN}
            width={NaN}
          />,
          <Handle
            innerRef={[Function]}
            left={NaN}
            onMouseDown={[Function]}
          />,
          <Handle
            innerRef={[Function]}
            left={NaN}
            onMouseDown={[Function]}
          />,
        ],
        "innerRef": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "class",
          "props": Object {
            "left": NaN,
            "width": NaN,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "class",
          "props": Object {
            "innerRef": [Function],
            "left": NaN,
            "onMouseDown": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "class",
          "props": Object {
            "innerRef": [Function],
            "left": NaN,
            "onMouseDown": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
